src/Core/Commit.js

```diff
@@ -0,0 +1,314 @@
+			switch (next.flag) {
+				case ElementComponent:
+					if (next.owner[LifecycleWillUnmount])
+						lifecycleMount(next, LifecycleWillUnmount)
+				default:
+					commitDetach(next, 1)
+					next = next.next
+			}
+	}
+
+	if (element.ref)
+		commitRef(element, element.ref, -1)
+
+	if (signature < 1) {
+		element.context = null
+		element.state = null
+		element.event = null
+		element.DOM = null
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {(function|string)?} callback
+ * @param {number} signature
+ * @param {*} key
+ */
+function commitRef (element, callback, signature, key) {
+	switch (typeof callback) {
+		case 'string':
+			return commitRef(element, componentRef, signature, callback)			
+		case 'function':
+			switch (signature) {
+				case -1:
+					return lifecycleCallback(element.host, callback, element.ref = null, key, element)
+				case 0:
+					element.ref = callback
+				case 1:
+					return lifecycleCallback(element.host, callback, element.instance || findDOMNode(element), key, element)
+				case 2:
+					commitRef(element, callback, -1, key)
+					commitRef(element, callback, 0, key)
+			}
+			break
+		default:
+			commitRef(element, element.ref || noop, -1, key)
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {string} type
+ * @param {(function|EventListener)} callback
+ */
+function commitEvent (element, type, callback) {
+	if (!element.event)
+		element.event = {}
+	
+	element.event[type] = callback
+	Event.addEventListener(element, type)
+}
+
+/**
+ * @param {Element} element
+ * @param {Object} props
+ * @param {number} signature
+ */
+function commitProps (element, props, signature) {
+	for (var key in props)
+		if (key === 'ref')
+			commitRef(element, props[key], signature)
+		else if (key.charCodeAt(0) === 111 && key.charCodeAt(1) === 110)
+			commitEvent(element, key.toLowerCase(), props[key])
+		else
+			DOMProperties(element, key, props[key], element.xmlns)
+}
+
+/**
+ * @param {Element} element
+ * @return {Object}
+ */
+function commitDOM (element) {
+	try {
+		return element.flag === ElementNode ? DOMElement(element) : DOMText(element)
+	} catch (e) {
+		return commitDOM(errorBoundary(element, e, LifecycleRender, 1))
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {(string|number)} value
+ */
+function commitValue (element, value) {
+	DOMValue(element, value)
+}
+
+/**
+ * @param {Element} element
+ */
+function commitContent (element) {
+	DOMContent(element)
+}
+
+/**
+ * @param {Element} element
+ * @param {Element} parent
+ */
+function commitRemove (element, parent) {
+	if (element.flag > ElementIntermediate)
+		DOMRemove(element, parent)
+	else
+		element.children.forEach(function (children) {
+			commitRemove(children, element.flag < ElementPortal ? parent : element)
+		})
+}
+
+/**
+ * @param {Element} element
+ * @param {Element} sibling
+ * @param {Element} parent
+ */
+function commitInsert (element, sibling, parent) {
+	if (sibling.flag < ElementIntermediate)
```

Comment:
comparison `<, >` of enums is quite difficult to follow

Reply:
There are 7 enum node type.

```
var ElementPromise = -3
var ElementFragment = -2
var ElementPortal = -1
var ElementIntermediate = 0
var ElementComponent = 1
var ElementNode = 2
var ElementText = 3
```
Everything less than `ElementIntermediate` is a compound node, everything greater than `ElementIntermediate` regular nodes.